src/ferny/ssh_askpass.py

```diff
@@ -0,0 +1,200 @@
+    # permitted (ie: the askpass callback should never return).
+    answers: ClassVar[Optional[Sequence[str]]] = None
+
+    # Patterns to capture.  `_pattern` *must* match.
+    _pattern: ClassVar[str]
+    # `_extra_patterns` can fill in extra class attributes if they match.
+    _extra_patterns: ClassVar[Sequence[str]] = ()
+
+    def __init__(self, prompt: str, messages: str, stderr: str, match: Match) -> None:
+        super().__init__(prompt, messages, stderr)
+        self.__dict__.update(match.groupdict())
+
+        for pattern in self._extra_patterns:
+            extra_match = re.search(with_helpers(pattern), messages, re.M)
+            print(extra_match, with_helpers(pattern), messages)
+            if extra_match is not None:
+                self.__dict__.update(extra_match.groupdict())
+
+
+# Specific prompts
+HELPERS = {
+    "%{algorithm}": r"(?P<algorithm>\b[-\w]+\b)",
+    "%{filename}": r"(?P<filename>.+)",
+    "%{fingerprint}": r"(?P<fingerprint>SHA256:[0-9A-Za-z+/]{43})",
+    "%{hostname}": r"(?P<hostname>[^ @']+)",
+    "%{pkcs11_id}": r"(?P<pkcs11_id>.+)",
+    "%{username}": r"(?P<username>[^ @']+)",
+}
+
+
+class PasswordPrompt(SSHAskpassPrompt):
+    _pattern = r"%{username}@%{hostname}'s password: "
+    username: Optional[str] = None
+    hostname: Optional[str] = None
+
+    async def dispatch(self, responder: 'SshAskpassResponder') -> Optional[str]:
+        return await responder.do_password_prompt(self)
+
+
+class PassphrasePrompt(SSHAskpassPrompt):
+    _pattern = r"Enter passphrase for key '%{filename}': "
+    filename: str
+
+    async def dispatch(self, responder: 'SshAskpassResponder') -> Optional[str]:
+        return await responder.do_passphrase_prompt(self)
+
+
+class FIDOPINPrompt(SSHAskpassPrompt):
+    _pattern = r"Enter PIN for %{algorithm} key %{filename}: "
+    algorithm: str
+    filename: str
+
+    async def dispatch(self, responder: 'SshAskpassResponder') -> Optional[str]:
+        return await responder.do_fido_pin_prompt(self)
+
+
+class FIDOUserPresencePrompt(SSHAskpassPrompt):
+    _pattern = r"Confirm user presence for key %{algorithm} %{fingerprint}\n"
+    answers = ()
+    algorithm: str
+    fingerprint: str
+
+    async def dispatch(self, responder: 'SshAskpassResponder') -> Optional[str]:
+        return await responder.do_fido_user_presence_prompt(self)
+
+
+class PKCS11PINPrompt(SSHAskpassPrompt):
+    _pattern = r"Enter PIN for '%{pkcs11_id}': "
+    pkcs11_id: str
+
+    async def dispatch(self, responder: 'SshAskpassResponder') -> Optional[str]:
+        return await responder.do_pkcs11_pin_prompt(self)
+
+
+class HostKeyPrompt(SSHAskpassPrompt):
+    _pattern = r"Are you sure you want to continue connecting \(yes/no(/\[fingerprint\])?\)\? "
+    _extra_patterns = [
+        r"%{fingerprint}[.]$",
+        r"^%{algorithm} key fingerprint is",
+        r"^The fingerprint for the %{algorithm} key sent by the remote host is$"
```

Comment:
I'm really nervous about such tight matches, assuming line breaks at specific places and such. But :man_shrugging: if it breaks, cockpit's tests should tell us. Between C8S and current Fedora we should also cover a wide range of OpenSSH versions.

Reply:
I agree that this is a bit dangerous, but the whole thing is a bit dangerous.  I'd rather find out about breaks.